In [1]:
import pandas as pd

pd.set_option('display.max_columns', 30)

# Invoice Generation from CUR

In [2]:
catalog.list()

['s3_pocs_cur_dataset',
 's3_intelbras_cur_dataset',
 's3_infra2_cur_dataset',
 'accounts_per_organization',
 'pocs_cur_dataset',
 'intelbras_cur_dataset',
 'infra2_cur_dataset',
 'processed_accounts_per_organization',
 'pocs_cur_dataset_merged',
 'intelbras_cur_dataset_merged',
 'infra2_cur_dataset_merged',
 'infra2_cur_dataset_agg',
 'parameters',
 'params:data_prep_raw_params',
 'params:data_prep_raw_params.kept_columns']

## Analyzing Infra 2 CUR of 2023-05

## Loading from S3

In [1]:
s3_infra2_cur = list(catalog.load('s3_infra2_cur_dataset').items())
s3_infra2_cur202305 = [t for t in s3_infra2_cur if t[0] == 'year=2023/month=5/CUR-00001.snappy.parquet'][0]
infra2_cur202305 = s3_infra2_cur202305[1]()
infra2_cur202305

[07/07/23 19:11:04] INFO     Loading data from 's3_infra2_cur_dataset' (PartitionedDataSet)...  ]8;id=246037;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=124185;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py#345\345]8;;\

identity_line_item_id  \
0     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
1     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
2     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
3     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
4     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
...                                                 ...   
5509  leyy3urvpooyfs3iai36yaekmq4vxyvzlqzyyk47um2mbo...   
5510  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   
5511  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   
5512  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   
5513  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   

                         identity_time_interval bill_invoice_id  \
0     2023-05-01T00:00:00Z/2023-05-01T01:00:00Z      1349042489   
1     2023-05-01T01:00:00Z/2023-05-01T02:00:00Z      1349042489   
2     2023-05-01T02:00:00Z/2023-05-01T03:00:00Z      1349042489   
3     2023-05-01T03:00:00Z/2023-05-01T04:00:00Z      1349042489   
4     2023-05-01T04:00:00Z/2023-05-01T05:00:00Z      1349042489   
...                                         ...             ...   
5509  2023-05-31T14:00:00Z/2023-05-31T15:00:00Z      1349042489   
5510  2023-05-31T15:00:00Z/2023-05-31T16:00:00Z      1349042489   
5511  2023-05-31T16:00:00Z/2023-05-31T17:00:00Z      1349042489   
5512  2023-05-31T17:00:00Z/2023-05-31T18:00:00Z      1349042489   
5513  2023-05-31T18:00:00Z/2023-05-31T19:00:00Z      1349042489   

          bill_invoicing_entity bill_billing_entity bill_bill_type  \
0     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
1     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
2     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
3     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
4     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
...                         ...                 ...            ...   
5509  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5510  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5511  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5512  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5513  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   

     bill_payer_account_id bill_billing_period_start_date  \
0             981064078967                     2023-05-01   
1             981064078967                     2023-05-01   
2             981064078967                     2023-05-01   
3             981064078967                     2023-05-01   
4             981064078967                     2023-05-01   
...                    ...                            ...   
5509          981064078967                     2023-05-01   
5510          981064078967                     2023-05-01   
5511          981064078967                     2023-05-01   
5512          981064078967                     2023-05-01   
5513          981064078967                     2023-05-01   

     bill_billing_period_end_date line_item_usage_account_id  ...  \
0                      2023-06-01               273762107511  ...   
1                      2023-06-01               273762107511  ...   
2                      2023-06-01               273762107511  ...   
3                      2023-06-01               273762107511  ...   
4                      2023-06-01               273762107511  ...   
...                           ...                        ...  ...   
5509                   2023-06-01               167002523613  ...   
5510                   2023-06-01               167002523613  ...   
5511                   2023-06-01               167002523613  ...   
5512                   2023-06-01               167002523613  ...   
5513                   2023-06-01               167002523613  ...   

                      resource_tags_aws_created_by  \
0                                Root:273762107511   
1        

In [2]:
[col for col in infra2_cur202305.columns if 'discount' in col]

['discount_spp_discount', 'discount_total_discount']

## Loading CUR1 from 01_raw

In [15]:
infra2_cur = list(catalog.load('infra2_cur_dataset').items())
infra2_cur

[07/06/23 18:49:34] INFO     Loading data from 'infra2_cur_dataset' (PartitionedDataSet)...     ]8;id=246454;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=639530;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py#345\345]8;;\

[('2023_05_CUR-00001.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd6206ad0>>),
 ('2023_05_CUR-00002.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bdc03fe80>>),
 ('2023_05_CUR-00003.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bdc03f460>>),
 ('2023_05_CUR-00004.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd598d7b0>>),
 ('2023_05_CUR-00005.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd59a60b0>>),
 ('2023_05_CUR-00006.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd59a5de0>>),
 ('2023_05_CUR-00007.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pa

In [16]:
infra2_cur1 = infra2_cur[0][1]()
infra2_cur1.head(100)

identity_line_item_id  bill_invoice_id  \
0   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
1   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
2   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
3   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
4   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
..                                                ...              ...   
95  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
96  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
97  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
98  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
99  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   

   bill_billing_entity      bill_invoicing_entity  \
0      AWS Marketplace  Amazon Web Services, Inc.   
1      AWS Marketplace  Amazon Web Services, Inc.   
2      AWS Marketplace  Amazon Web Services, Inc.   
3      AWS Marketplace  Amazon Web Services, Inc.   
4      AWS Marketplace  Amazon Web Services, Inc.   
..                 ...                        ...   
95     AWS Marketplace  Amazon Web Services, Inc.   
96     AWS Marketplace  Amazon Web Services, Inc.   
97     AWS Marketplace  Amazon Web Services, Inc.   
98     AWS Marketplace  Amazon Web Services, Inc.   
99     AWS Marketplace  Amazon Web Services, Inc.   

               line_item_legal_entity bill_payer_account_id  \
0   Cloud Infrastructure Services Ltd          981064078967   
1   Cloud Infrastructure Services Ltd          981064078967   
2   Cloud Infrastructure Services Ltd          981064078967   
3   Cloud Infrastructure Services Ltd          981064078967   
4   Cloud Infrastructure Services Ltd          981064078967   
..                                ...                   ...   
95  Cloud Infrastructure Services Ltd          981064078967   
96  Cloud Infrastructure Services Ltd          981064078967   
97  Cloud Infrastructure Services Ltd          981064078967   
98  Cloud Infrastructure Services Ltd          981064078967   
99  Cloud Infrastructure Services Ltd          981064078967   

   line_item_usage_account_id line_item_line_item_type  \
0                273762107511                    Usage   
1                273762107511                    Usage   
2                273762107511                    Usage   
3                273762107511                    Usage   
4                273762107511                    Usage   
..                        ...                      ...   
95               273762107511                    Usage   
96               273762107511                    Usage   
97               273762107511                    Usage   
98               273762107511                    Usage   
99               273762107511                    Usage   

   bill_billing_period_start_date bill_billing_period_end_date  \
0                      2023-05-01                   2023-06-01   
1                      2023-05-01                   2023-06-01   
2                      2023-05-01                   2023-06-01   
3                      2023-05-01                   2023-06-01   
4                      2023-05-01                   2023-06-01   
..                            ...                          ...   
95                     2023-05-01                   2023-06-01   
96                     2023-05-01                   2023-06-01   
97                     2023-05-01                   2023-06-01   
98                     2023-05-01                   2023-06-01   
99                     2023-05-01                   2023-06-01   

   line_item_usage_start_date line_item_usage_end_date  \
0         2023-05-01 00:00:00      2023-05-01 01:00:00   
1         2023-05-01 01:00:00      2023-05-01 02:00:00   
2         2023-05-01 02:00:00      2023-05-01 03:00:00   
3         2023-05-01 03:00:

In [6]:
infra2_cur1['line_item_usage_account_id'].dtype

dtype('O')

In [7]:
infra2_cur1['product_product_name'].unique()

array(['WordPress on Server 2019 + PhpMyAdmin + MySQL',
       'OpenVPN Access Server (5 Connected Devices)',
       'Oracle Linux 8 with Support by Supported Images',
       'ZoneMinder:  Build Your Own Security Camera Solution',
       'Oracle Linux 7 with Support by Supported Images',
       'Reverse Proxy Server using NGINX on Ubuntu',
       'Veeam Backup for AWS Paid Edition'], dtype=object)

## Replicating Infra 2 - Marketplace - Invoice ID 1349042489

### Detail for Consolidated Bill

In [18]:
infra2_cur1.groupby(by='product_product_name')['line_item_unblended_cost'].sum().reset_index()

product_product_name  line_item_unblended_cost
0        OpenVPN Access Server (5 Connected Devices)                   142.848
1    Oracle Linux 7 with Support by Supported Images                   147.100
2    Oracle Linux 8 with Support by Supported Images                   260.900
3         Reverse Proxy Server using NGINX on Ubuntu                     0.144
4                  Veeam Backup for AWS Paid Edition                    24.310
5      WordPress on Server 2019 + PhpMyAdmin + MySQL                    22.260
6  ZoneMinder:  Build Your Own Security Camera So...                     4.470

In [19]:
infra2_cur1['line_item_unblended_cost'].sum()

602.0319999999999

#### <font color='green'>Matches the PDF invoice</font>

### Linked Account Allocation
#### Activity By Account

In [20]:
infra2_cur1.groupby(by='line_item_usage_account_id')['line_item_unblended_cost'].sum()

line_item_usage_account_id
065716851408     73.900
167002523613    358.410
273762107511     22.260
465517541486     71.424
563324065975     71.424
901519325564      4.614
Name: line_item_unblended_cost, dtype: float64

In [21]:
len('065716851408'), len('167002523613'), len('273762107511'), len('465517541486'), len('563324065975'), len('901519325564')

(12, 12, 12, 12, 12, 12)

#### <font color='green'>Matches the PDF invoice</font>

In [22]:
infra2_cur1.groupby(by=['line_item_usage_account_id', 'product_product_name'])['line_item_unblended_cost'].sum()

line_item_usage_account_id  product_product_name                                
065716851408                Oracle Linux 8 with Support by Supported Images          73.900
167002523613                Oracle Linux 7 with Support by Supported Images         147.100
                            Oracle Linux 8 with Support by Supported Images         187.000
                            Veeam Backup for AWS Paid Edition                        24.310
273762107511                WordPress on Server 2019 + PhpMyAdmin + MySQL            22.260
465517541486                OpenVPN Access Server (5 Connected Devices)              71.424
563324065975                OpenVPN Access Server (5 Connected Devices)              71.424
901519325564                Reverse Proxy Server using NGINX on Ubuntu                0.144
                            ZoneMinder:  Build Your Own Security Camera Solution      4.470
Name: line_item_unblended_cost, dtype: float64

## Loading CUR2 from 01_raw

In [23]:
infra2_cur2 = infra2_cur[1][1]()
infra2_cur2.head(100)

identity_line_item_id  bill_invoice_id  \
0  t5fvp5rwqyrvscf3kc2ylwmobdyaazby3yjrbuz4gslxvl...       1366015353   
1  nmuam35xnovmsc7ybj44am7r575ul3obphzy6xln66jykt...       1366015353   
2  kieg2eczb3cyj665s4hklsiw64xx6ldkzl34jgamkyuodv...       1366015353   
3  ikif2qfz7mzuwkwgilejfosjahuhoh5tjhpnrg2uecgan5...       1366015353   

  bill_billing_entity             bill_invoicing_entity  \
0                 AWS  Amazon AWS Serviços Brasil Ltda.   
1                 AWS  Amazon AWS Serviços Brasil Ltda.   
2                 AWS  Amazon AWS Serviços Brasil Ltda.   
3                 AWS  Amazon AWS Serviços Brasil Ltda.   

             line_item_legal_entity bill_payer_account_id  \
0  Amazon AWS Serviços Brasil Ltda.          981064078967   
1  Amazon AWS Serviços Brasil Ltda.          981064078967   
2  Amazon AWS Serviços Brasil Ltda.          981064078967   
3  Amazon AWS Serviços Brasil Ltda.          981064078967   

  line_item_usage_account_id line_item_line_item_type  \
0               072038246234                   Refund   
1               072038246234                      Tax   
2               562293755332                   Refund   
3               562293755332                      Tax   

  bill_billing_period_start_date bill_billing_period_end_date  \
0                     2023-05-01                   2023-06-01   
1                     2023-05-01                   2023-06-01   
2                     2023-05-01                   2023-06-01   
3                     2023-05-01                   2023-06-01   

  line_item_usage_start_date line_item_usage_end_date  line_item_usage_amount  \
0                 2023-05-01               2023-06-01                     0.0   
1                 2023-05-01               2023-06-01                     1.0   
2                 2023-05-01               2023-06-01                     0.0   
3                 2023-05-01               2023-06-01                     1.0   

  line_item_product_code    product_product_name line_item_usage_type  \
0     AWSSupportBusiness  AWS Support (Business)               Dollar   
1     AWSSupportBusiness  AWS Support (Business)               Dollar   
2     AWSSupportBusiness  AWS Support (Business)                  NaN   
3     AWSSupportBusiness  AWS Support (Business)                  NaN   

   line_item_operation  product_from_location_type  product_from_region_code  \
0                  NaN                         NaN                       NaN   
1                  NaN                         NaN                       NaN   
2                  NaN                         NaN                       NaN   
3                  NaN                         NaN                       NaN   

  line_item_tax_type line_item_currency_code  line_item_unblended_cost  
0                NaN                     USD                    -57.61  
1                Tax                     USD                     -7.97  
2                NaN                     USD                     -7.00  
3                Tax                     USD                     -0.96

In [24]:
infra2_cur2['line_item_unblended_cost'].sum()

-73.53999999999999

## Loading CUR3 from 01_raw

In [27]:
infra2_cur3 = infra2_cur[2][1]()
print(infra2_cur3.shape)
infra2_cur3.head(100)

[07/06/23 19:13:12] WARNING  /home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-pa ]8;id=1108;file:///usr/lib/python3.10/warnings.py\warnings.py]8;;\:]8;id=942890;file:///usr/lib/python3.10/warnings.py#109\109]8;;\
                             ckages/kedro_datasets/pandas/csv_dataset.py:162: DtypeWarning: Columns                
                             (19) have mixed types. Specify dtype option on import or set                          
                             low_memory=False.                                                                     
                               return pd.read_csv(load_path, **self._load_args)                                    
                                                                                                                   

(764846, 22)


identity_line_item_id  bill_invoice_id  \
0   qmgjvd6gxycmgwrk62rqqkjr63ex5ghqoafbqzurlrpc24...       1343605689   
1   tkd3ppomattn3wyjiak4bp6zlmlo3uaffrnklmk4b3fwdi...       1343605689   
2   fd6dcj5scxyopuenl3n2ighvic2nhuaz3i4fhpanbr3bqf...       1343605689   
3   xldj5f3wd2b4hzzok7zujmonpgcra6uumx4vlvdr6oig7a...       1343605689   
4   t3msytn2lq6tilz4y3pxmqqqqyjhhifkk6qcpkinq2ludy...       1343605689   
..                                                ...              ...   
95  haixin2uwlwnlm5v4ucldjjqrrspevkdnshn4gwewfdffc...       1343605689   
96  pe5a67pil3j22kxjfmztbknf6yquq2kijhvg5bmsrxplba...       1343605689   
97  yg3gumehs3htpiu33fzt5es2wvin2hxjoz2u35kps2tduh...       1343605689   
98  kc2tfbtef7gwxc6sf4f3vl2xq3sdu7gvhzenapvs5edadq...       1343605689   
99  hfyntlujo4vaxvmdnph45dmrxecbpopmkt4ysjqipyh6it...       1343605689   

   bill_billing_entity             bill_invoicing_entity  \
0                  AWS  Amazon AWS Serviços Brasil Ltda.   
1                  AWS  Amazon AWS Serviços Brasil Ltda.   
2                  AWS  Amazon AWS Serviços Brasil Ltda.   
3                  AWS  Amazon AWS Serviços Brasil Ltda.   
4                  AWS  Amazon AWS Serviços Brasil Ltda.   
..                 ...                               ...   
95                 AWS  Amazon AWS Serviços Brasil Ltda.   
96                 AWS  Amazon AWS Serviços Brasil Ltda.   
97                 AWS  Amazon AWS Serviços Brasil Ltda.   
98                 AWS  Amazon AWS Serviços Brasil Ltda.   
99                 AWS  Amazon AWS Serviços Brasil Ltda.   

              line_item_legal_entity bill_payer_account_id  \
0   Amazon AWS Serviços Brasil Ltda.          981064078967   
1   Amazon AWS Serviços Brasil Ltda.          981064078967   
2   Amazon AWS Serviços Brasil Ltda.          981064078967   
3   Amazon AWS Serviços Brasil Ltda.          981064078967   
4   Amazon AWS Serviços Brasil Ltda.          981064078967   
..                               ...                   ...   
95  Amazon AWS Serviços Brasil Ltda.          981064078967   
96  Amazon AWS Serviços Brasil Ltda.          981064078967   
97  Amazon AWS Serviços Brasil Ltda.          981064078967   
98  Amazon AWS Serviços Brasil Ltda.          981064078967   
99  Amazon AWS Serviços Brasil Ltda.          981064078967   

   line_item_usage_account_id line_item_line_item_type  \
0                072038246234                    Usage   
1                368505959464                    Usage   
2                543280990914                    Usage   
3                368505959464                    Usage   
4                420065387364                    Usage   
..                        ...                      ...   
95               901519325564                    Usage   
96               684481376085                    Usage   
97               901519325564                    Usage   
98               494939165758                    Usage   
99               072038246234                    Usage   

   bill_billing_period_start_date bill_billing_period_end_date  \
0                      2023-05-01                   2023-06-01   
1                      2023-05-01                   2023-06-01   
2                      2023-05-01                   2023-06-01   
3                      2023-05-01                   2023-06-01   
4                      2023-05-01                   2023-06-01   
..                            ...                          ...   
95                     2023-05-01                   2023-06-01   
96                     2023-05-01                   2023-06-01   
97                     2023-05-01                   2023-06-01   
98                     2023-05-01                   2023-06-01   
99                     2023-05-01                   2023-06-01   

   line_item_usage_start_date line_item_usage_end_date  \
0         2023-05-15 16:00:00      2023-05-15 17:00:00   
1         2023-05-25 20:00:00      2023-05-25 21:00:00   
2         202

In [28]:
infra2_cur3.groupby(by=['line_item_usage_account_id', 'product_product_name', 'line_item_currency_code'])['line_item_unblended_cost'].sum()

line_item_usage_account_id  product_product_name                line_item_currency_code
065716851408                AWS Config                          USD                          0.282000
                            AWS Key Management Service          USD                          0.000150
                            Amazon Elastic Compute Cloud        USD                        197.672197
                            Amazon Simple Notification Service  USD                          0.000001
                            Amazon Simple Queue Service         USD                          0.000004
                                                                                              ...    
988350742118                Amazon Elastic Compute Cloud        USD                         11.680824
                            Amazon Relational Database Service  USD                         18.545119
                            Amazon Simple Storage Service       USD                          7.2

In [29]:
a = pd.DataFrame({'a': [1, 2, 3], 'b': [1, 2, 3]})
b = pd.DataFrame({'a': [4, 5, 6], 'b': [4, 5, 6]})

pd.concat([a,b])

a  b
0  1  1
1  2  2
2  3  3
0  4  4
1  5  5
2  6  6

# Aggregating costs by invoice, account and product

In [1]:
catalog.list()

['s3_pocs_cur_dataset',
 's3_intelbras_cur_dataset',
 's3_infra2_cur_dataset',
 'accounts_per_organization',
 'pocs_cur_dataset',
 'intelbras_cur_dataset',
 'infra2_cur_dataset',
 'processed_accounts_per_organization',
 'pocs_cur_dataset_merged',
 'intelbras_cur_dataset_merged',
 'infra2_cur_dataset_merged',
 'infra2_cur_dataset_agg',
 'parameters',
 'params:data_prep_raw_params',
 'params:data_prep_raw_params.kept_columns']

In [3]:
df_cur = catalog.load('infra2_cur_dataset_merged')
df_cur

[07/12/23 14:34:25] INFO     Loading data from 'infra2_cur_dataset_merged' (CSVDataSet)...      ]8;id=586379;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=907059;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py#345\345]8;;\

[07/12/23 14:34:55] WARNING  /home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-pa ]8;id=90265;file:///usr/lib/python3.10/warnings.py\warnings.py]8;;\:]8;id=429936;file:///usr/lib/python3.10/warnings.py#109\109]8;;\
                             ckages/kedro_datasets/pandas/csv_dataset.py:162: DtypeWarning: Columns                
                             (4,14,17,18,19) have mixed types. Specify dtype option on import or                   
                             set low_memory=False.                                                                 
                               return pd.read_csv(load_path, **self._load_args)                                    
                                                                                                                   

identity_line_item_id  bill_invoice_id  \
0        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
1        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
2        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
3        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
4        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
...                                                    ...              ...   
3835113  gbp2uwf4otex4r3i6x2xaqmecr4zo73kcdsjvjaci3cdew...       1343605689   
3835114  7lfdzbjdj5foh27w3ht5rgu32dkugujmsx6krg22hqyn5c...       1343605689   
3835115  gbp2uwf4otex4r3i6x2xaqmecr4zo73kcdsjvjaci3cdew...       1343605689   
3835116  7lfdzbjdj5foh27w3ht5rgu32dkugujmsx6krg22hqyn5c...       1343605689   
3835117  gbp2uwf4otex4r3i6x2xaqmecr4zo73kcdsjvjaci3cdew...       1343605689   

        bill_billing_entity             bill_invoicing_entity  \
0           AWS Marketplace         Amazon Web Services, Inc.   
1           AWS Marketplace         Amazon Web Services, Inc.   
2           AWS Marketplace         Amazon Web Services, Inc.   
3           AWS Marketplace         Amazon Web Services, Inc.   
4           AWS Marketplace         Amazon Web Services, Inc.   
...                     ...                               ...   
3835113                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835114                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835115                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835116                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835117                 AWS  Amazon AWS Serviços Brasil Ltda.   

                    line_item_legal_entity bill_payer_account_id  \
0        Cloud Infrastructure Services Ltd          981064078967   
1        Cloud Infrastructure Services Ltd          981064078967   
2        Cloud Infrastructure Services Ltd          981064078967   
3        Cloud Infrastructure Services Ltd          981064078967   
4        Cloud Infrastructure Services Ltd          981064078967   
...                                    ...                   ...   
3835113                                NaN          981064078967   
3835114                                NaN          981064078967   
3835115                                NaN          981064078967   
3835116                                NaN          981064078967   
3835117                                NaN          981064078967   

        line_item_usage_account_id line_item_line_item_type  \
0                     273762107511                    Usage   
1                     273762107511                    Usage   
2                     273762107511                    Usage   
3                     273762107511                    Usage   
4                     273762107511                    Usage   
...                            ...                      ...   
3835113               543280990914              SppDiscount   
3835114               637926748998              SppDiscount   
3835115               543280990914              SppDiscount   
3835116               637926748998              SppDiscount   
3835117               543280990914              SppDiscount   

        bill_billing_period_start_date bill_billing_period_end_date  \
0                           2023-05-01                   2023-06-01   
1                           2023-05-01                   2023-06-01   
2                           2023-05-01                   2023-06-01   
3                           2023-05-01                   2023-06-01   
4                           2023-05-01                   2023-06-01   
...                                ...                          ...   
3835113                     2023-05-01                   2023-06-01   
3835114                     2023-05-01                   2023-06-01   
3835115                     2023-05-01                   2023-06-01   
3835116                

In [3]:
df_accounts_per_org = catalog.load('processed_accounts_per_organization')
df_accounts_per_org

[07/06/23 21:50:58] INFO     Loading data from 'processed_accounts_per_organization'            ]8;id=684332;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=174320;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py#345\345]8;;\
                             (CSVDataSet)...                                                                       

account_id                                    name
0   705000000000                    FinOps Processamento
1   594000000000  SES - Secretaria da Saude - Assessment
2   895000000000               Intelbras - Container PoC
3   665000000000                      Disc Day - Estacio
4   609000000000                      ALESP - Assessment
..           ...                                     ...
63  564000000000                                   Shift
64  400000000000                                  TRN GB
65  913000000000                                 GB TRN2
66  623000000000              intelbras-teste-networking
67  550000000000                  intelbras-teste-spoke1

[68 rows x 2 columns]

In [9]:
import logging

logger = logging.getLogger()

In [10]:
import pandas as pd

In [11]:
def add_account_names(df_cur: pd.DataFrame, df_accounts_per_org: pd.DataFrame) -> pd.DataFrame:

    logger.info('Adding account names...')
    df_cur = df_cur.merge(df_accounts_per_org, how='left', left_on='bill_payer_account_id', right_on='account_id', suffixes=('', '_payer'), validate='many_to_one')
    df_cur = df_cur.rename(columns={'name': 'payer_account_name'})
    df_cur = df_cur.merge(df_accounts_per_org, how='left', left_on='line_item_usage_account_id', right_on='account_id', suffixes=('', '_usage'), validate='many_to_one')
    df_cur = df_cur.rename(columns={'name': 'usage_account_name'})

    return df_cur

df_cur_enriched = add_account_names(df_cur, df_accounts_per_org)
df_cur_enriched

identity_line_item_id  bill_invoice_id  \
0        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
1        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
2        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
3        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
4        xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
...                                                    ...              ...   
3835113  gbp2uwf4otex4r3i6x2xaqmecr4zo73kcdsjvjaci3cdew...       1343605689   
3835114  7lfdzbjdj5foh27w3ht5rgu32dkugujmsx6krg22hqyn5c...       1343605689   
3835115  gbp2uwf4otex4r3i6x2xaqmecr4zo73kcdsjvjaci3cdew...       1343605689   
3835116  7lfdzbjdj5foh27w3ht5rgu32dkugujmsx6krg22hqyn5c...       1343605689   
3835117  gbp2uwf4otex4r3i6x2xaqmecr4zo73kcdsjvjaci3cdew...       1343605689   

        bill_billing_entity             bill_invoicing_entity  \
0           AWS Marketplace         Amazon Web Services, Inc.   
1           AWS Marketplace         Amazon Web Services, Inc.   
2           AWS Marketplace         Amazon Web Services, Inc.   
3           AWS Marketplace         Amazon Web Services, Inc.   
4           AWS Marketplace         Amazon Web Services, Inc.   
...                     ...                               ...   
3835113                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835114                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835115                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835116                 AWS  Amazon AWS Serviços Brasil Ltda.   
3835117                 AWS  Amazon AWS Serviços Brasil Ltda.   

                    line_item_legal_entity bill_payer_account_id  \
0        Cloud Infrastructure Services Ltd          981064078967   
1        Cloud Infrastructure Services Ltd          981064078967   
2        Cloud Infrastructure Services Ltd          981064078967   
3        Cloud Infrastructure Services Ltd          981064078967   
4        Cloud Infrastructure Services Ltd          981064078967   
...                                    ...                   ...   
3835113                                NaN          981064078967   
3835114                                NaN          981064078967   
3835115                                NaN          981064078967   
3835116                                NaN          981064078967   
3835117                                NaN          981064078967   

        line_item_usage_account_id line_item_line_item_type  \
0                     273762107511                    Usage   
1                     273762107511                    Usage   
2                     273762107511                    Usage   
3                     273762107511                    Usage   
4                     273762107511                    Usage   
...                            ...                      ...   
3835113               543280990914              SppDiscount   
3835114               637926748998              SppDiscount   
3835115               543280990914              SppDiscount   
3835116               637926748998              SppDiscount   
3835117               543280990914              SppDiscount   

        bill_billing_period_start_date bill_billing_period_end_date  ...  \
0                           2023-05-01                   2023-06-01  ...   
1                           2023-05-01                   2023-06-01  ...   
2                           2023-05-01                   2023-06-01  ...   
3                           2023-05-01                   2023-06-01  ...   
4                           2023-05-01                   2023-06-01  ...   
...                                ...                          ...  ...   
3835113                     2023-05-01                   2023-06-01  ...   
3835114                     2023-05-01                   2023-06-01  ...   
3835115                     2023-05-01           

In [4]:
df_cur_merged = df_cur_enriched.groupby(by=['bill_invoice_id', 'bill_billing_entity', 'bill_invoicing_entity', 
                                          'bill_payer_account_id', 'line_item_usage_account_id',
                                          'line_item_line_item_type', 'line_item_usage_start_date', 
                                          'line_item_usage_end_date', 'product_product_name', 'line_item_currency_code'])['line_item_unblended_cost'].sum()

df_cur_merged


╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:1                                                                                    │
│                                                                                                  │
│ ❱ 1 df_cur_merged = df_cur_enriched.groupby(by=['bill_invoice_id', 'bill_billing_entity', 'b     │
│   2 │   │   │   │   │   │   │   │   │   │     'bill_payer_account_id', 'line_item_usage_acco     │
│   3 │   │   │   │   │   │   │   │   │   │     'line_item_line_item_type', 'line_item_usage_s     │
│   4 │   │   │   │   │   │   │   │   │   │     'line_item_usage_end_date', 'product_product_n     │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
NameError: name 'df_cur_enriched' is not defined

# Generating Accounts Per Org file

In [28]:
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 100)
df = pd.read_csv('../data/01_raw/infra2_invoice.csv', dtype={'PayerAccountId': str, 'LinkedAccountId': str})
df.head(5)

InvoiceID PayerAccountId  payer_account_id  linked_account_id  \
0  1.349042e+09   981064078967      981064078967                  0   
1  1.349042e+09   981064078967      981064078967                  0   
2  1.349042e+09   981064078967      981064078967                  0   
3  1.349042e+09   981064078967      981064078967                  0   
4  1.349042e+09   981064078967      981064078967                  0   

  LinkedAccountId     RecordType RecordID BillingPeriodStartDate  \
0             NaN  PayerLineItem       -1          5/1/2023 0:00   
1             NaN  PayerLineItem  6.8E+18          5/1/2023 0:00   
2             NaN  PayerLineItem  6.8E+18          5/1/2023 0:00   
3             NaN  PayerLineItem  6.8E+18          5/1/2023 0:00   
4             NaN  PayerLineItem  6.8E+18          5/1/2023 0:00   

  BillingPeriodEndDate     InvoiceDate   PayerAccountName LinkedAccountName  \
0      5/31/2023 23:59  6/2/2023 16:47  dataRain Payeer 2               NaN   
1      5/31/2023 23:59  6/2/2023 16:47  dataRain Payeer 2               NaN   
2      5/31/2023 23:59  6/2/2023 16:47  dataRain Payeer 2               NaN   
3      5/31/2023 23:59  6/2/2023 16:47  dataRain Payeer 2               NaN   
4      5/31/2023 23:59  6/2/2023 16:47  dataRain Payeer 2               NaN   

   TaxationAddress  PayerPONumber                ProductCode  \
0              NaN            NaN  3ihdqli79gl9v2jnlzs6nq60h   
1              NaN            NaN  eryhlhzpi9qp90nh5u371dw9z   
2              NaN            NaN  eryhlhzpi9qp90nh5u371dw9z   
3              NaN            NaN  eryhlhzpi9qp90nh5u371dw9z   
4              NaN            NaN  625odl4lkfwv8zdsyfqsib0wy   

                                       ProductName  \
0      OpenVPN Access Server (5 Connected Devices)   
1  Oracle Linux 8 with Support by Supported Images   
2  Oracle Linux 8 with Support by Supported Images   
3  Oracle Linux 8 with Support by Supported Images   
4  Oracle Linux 7 with Support by Supported Images   

               SellerOfRecord                     UsageType Operation  \
0  OpenVPN Technologies, Inc.       SoftwareUsage:t3a.small    Hourly   
1            Supported Images  SAE1-SoftwareUsage:t2.medium    Hourly   
2            Supported Images  SAE1-SoftwareUsage:t3.xlarge    Hourly   
3            Supported Images   SAE1-SoftwareUsage:c5.large    Hourly   
4            Supported Images  SAE1-SoftwareUsage:t3.xlarge    Hourly   

         RateId                                    ItemDescription  \
0  1.719807e+10  AWS Marketplace hourly software usage|us-east-...   
1  1.918718e+10  AWS Marketplace hourly software usage|sa-east-...   
2  1.918717e+10  AWS Marketplace hourly software usage|sa-east-...   
3  1.918717e+10  AWS Marketplace hourly software usage|sa-east-...   
4  1.903625e+10  AWS Marketplace hourly software usage|sa-east-...   

  UsageStartDate     UsageEndDate  UsageQuantity  BlendedRate CurrencyCode  \
0  5/1/2023 0:00  5/31/2023 23:59         1488.0          NaN          USD   
1  5/1/2023 0:00  5/31/2023 23:59          739.0          NaN          USD   
2  5/1/2023 0:00  5/31/2023 23:59           10.0          NaN          USD   
3  5/1/2023 0:00  5/31/2023 23:59            6.0          NaN          USD   
4  5/1/2023 0:00  5/31/2023 23:59          733.0          NaN          USD   

   CostBeforeTax  Credits  TaxAmount TaxType  TotalCost  981064078967  
0         142.84      0.0        0.0     NaN     142.84           NaN  
1          73.90      0.0        0.0     NaN      73.90           NaN  
2           2.00      0.0        0.0     NaN       2.00           NaN  
3           0.60      0.0        0.0     NaN       0.60           NaN  
4         146.60      0.0        0.0     NaN     146.60           NaN

In [37]:
df_payer = df[['payer_account_id', 'PayerAccountName']].copy()
df_payer = df_payer.drop_duplicates()
df_payer = df_payer.rename(columns={'payer_account_id': 'Account ID', 'PayerAccountName': 'Name'})
df_payer

Account ID               Name
0  981064078967  dataRain Payeer 2

In [38]:
df_linked = df[['linked_account_id', 'LinkedAccountName']].copy()
df_linked = df_linked.drop_duplicates().dropna()
df_linked = df_linked.rename(columns={'linked_account_id': 'Account ID', 'LinkedAccountName': 'Name'})
df_linked

Account ID                                     Name
11     65716851408      SPI - SP Intervention - ENE Medical
12    167002523613                    SPI - SP Intervention
18    273762107511                                 Cordvida
19    465517541486                                Bipp Care
20    563324065975                               Symbiomics
21    901519325564                          Estre Ambiental
1172   72038246234                                     SMIT
1474   95507049355                                     IHSL
1532  113082347426                          Estre - Conta 2
1673  215028743710                          Hospital Sabara
1684  244835640994                            GTM Chemicals
1803  350427820123                                   Oliv-e
1822  368505959464  [DEV] OTIMO SOCIEDADE DE CREDITO DIRETO
2095  415870527538                      Hospital Santa Cruz
2178  420065387364                                   pakman
2454  435585749363                                      EOS
2497  446999942749                           Ceramica Atlas
2602  468681385603                  Treinamento VentureLabs
2605  494939165758                             [PRD] Helpie
2965  511871492403                          [PRD] Diligente
3063  543280990914  [PRD] OTIMO SOCIEDADE DE CREDITO DIRETO
3357  562293755332                                 Cotrijal
3593  591891948593                     Producao VentureLabs
3595  623041266269                             PrimeDoctors
3637  637926748998                             [DEV] Helpie
3809  684481376085                                   Taksim
3951  702894359990                             PUC-Campinas
4022  752368805386                            [PRD] BluePay
4055  802316276539                            libelulanegra
4250  940848700982                                  HCFMUSP
4309  946577008532                                    JR_GX
4310  981064078967                        dataRain Payeer 2
4327  988350742118                                     IHSL

In [17]:
df_accounts_per_org = pd.concat([df_payer, df_linked]).drop_duplicates()
df_accounts_per_org

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:1                                                                                    │
│                                                                                                  │
│ ❱ 1 df_accounts_per_org = pd.concat([df_payer, df_linked]).drop_duplicates()                     │
│   2 df_accounts_per_org                                                                          │
│   3                                                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
NameError: name 'df_payer' is not defined

In [42]:
df_accounts_per_org.to_csv('../data/01_raw/accounts_per_organization.csv', index=False)

## Aggregated Infra2 Invoices

In [1]:
catalog.list()

['s3_pocs_cur_dataset',
 's3_intelbras_cur_dataset',
 's3_infra2_cur_dataset',
 'accounts_per_organization',
 'pocs_cur_dataset',
 'intelbras_cur_dataset',
 'infra2_cur_dataset',
 'processed_accounts_per_organization',
 'pocs_cur_dataset_merged',
 'intelbras_cur_dataset_merged',
 'infra2_cur_dataset_merged',
 'infra2_cur_dataset_agg',
 'parameters',
 'params:data_prep_raw_params',
 'params:data_prep_raw_params.kept_columns']

In [18]:
df_infra2_agg = catalog.load('infra2_cur_dataset_agg')
df_infra2_agg

[07/12/23 17:58:14] INFO     Loading data from 'infra2_cur_dataset_agg' (CSVDataSet)...         ]8;id=616255;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=429040;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py#345\345]8;;\

bill_invoice_id bill_billing_entity             bill_invoicing_entity  \
0            1343605689                 AWS  Amazon AWS Serviços Brasil Ltda.   
1            1343605689                 AWS  Amazon AWS Serviços Brasil Ltda.   
2            1343605689                 AWS  Amazon AWS Serviços Brasil Ltda.   
3            1343605689                 AWS  Amazon AWS Serviços Brasil Ltda.   
4            1343605689                 AWS  Amazon AWS Serviços Brasil Ltda.   
...                 ...                 ...                               ...   
183519       1349042489     AWS Marketplace         Amazon Web Services, Inc.   
183520       1366015353                 AWS  Amazon AWS Serviços Brasil Ltda.   
183521       1366015353                 AWS  Amazon AWS Serviços Brasil Ltda.   
183522       1366015353                 AWS  Amazon AWS Serviços Brasil Ltda.   
183523       1366015353                 AWS  Amazon AWS Serviços Brasil Ltda.   

       bill_payer_account_id payer_account_name line_item_usage_account_id  \
0               981064078967  dataRain Payeer 2               065716851408   
1               981064078967  dataRain Payeer 2               065716851408   
2               981064078967  dataRain Payeer 2               065716851408   
3               981064078967  dataRain Payeer 2               065716851408   
4               981064078967  dataRain Payeer 2               065716851408   
...                      ...                ...                        ...   
183519          981064078967  dataRain Payeer 2               901519325564   
183520          981064078967  dataRain Payeer 2               072038246234   
183521          981064078967  dataRain Payeer 2               072038246234   
183522          981064078967  dataRain Payeer 2               562293755332   
183523          981064078967  dataRain Payeer 2               562293755332   

                         usage_account_name line_item_line_item_type  \
0       SPI - SP Intervention - ENE Medical                      Tax   
1       SPI - SP Intervention - ENE Medical                      Tax   
2       SPI - SP Intervention - ENE Medical                      Tax   
3       SPI - SP Intervention - ENE Medical                      Tax   
4       SPI - SP Intervention - ENE Medical                      Tax   
...                                     ...                      ...   
183519                      Estre Ambiental                    Usage   
183520                                 SMIT                   Refund   
183521                                 SMIT                      Tax   
183522                             Cotrijal                   Refund   
183523                             Cotrijal                      Tax   

       line_item_usage_start_date line_item_usage_end_date  \
0             2023-05-01 00:00:00      2023-06-01 00:00:00   
1             2023-05-01 00:00:00      2023-06-01 00:00:00   
2             2023-05-01 00:00:00      2023-06-01 00:00:00   
3             2023-05-01 00:00:00      2023-06-01 00:00:00   
4             2023-05-01 00:00:00      2023-06-01 00:00:00   
...                           ...                      ...   
183519        2023-05-31 23:00:00      2023-06-01 00:00:00   
183520                 2023-05-01               2023-06-01   
183521                 2023-05-01               2023-06-01   
183522                 2023-05-01               2023-06-01   
183523                 2023-05-01               2023-06-01   

                                     product_product_name  \
0                                              AWS Config   
1                                       AWS Data Transfer   
2                              AWS Key Management Service   
3                            Amazon Elastic Compute Cloud   
4                      Amazon Simple Notification Service   
...                                                   ...   
183519  ZoneMinder:  Build Your Own Security Camera So...   


In [13]:
'065716851408' in df_infra2_agg['line_item_usage_account_id'].unique()

False

In [15]:
'072038246234' in df_infra2_agg['line_item_usage_account_id'].unique()

False

In [19]:
df_infra2_agg['bill_billing_entity'].unique()

array(['AWS', 'AWS Marketplace'], dtype=object)

In [20]:
df_infra2_agg['bill_invoice_id'].unique()

array([1343605689, 1349042489, 1366015353])

In [21]:
df_infra2_agg['line_item_line_item_type'].unique()

array(['Tax', 'Usage', 'Credit', 'Fee', 'Refund'], dtype=object)

In [30]:
df_consume = df_infra2_agg[df_infra2_agg['bill_invoice_id']==1343605689]
df_consume_usage = df_consume[df_consume['line_item_line_item_type']=='Usage']
print(f"Consume charges total: {df_consume_usage['line_item_unblended_cost'].sum()}")

Consume charges total: 32841.2004781926


In [29]:
df_consume = df_infra2_agg[df_infra2_agg['bill_invoice_id']==1343605689]
df_consume_tax = df_consume[df_consume['line_item_line_item_type']=='Tax']
print(f"Consume tax total: {df_consume_tax['line_item_unblended_cost'].sum()}")

Consume tax total: 3610.5000000000005


In [28]:
spp_discount = df_infra2_agg[df_infra2_agg['bill_invoice_id']==1343605689]['discount_spp_discount'].sum()
print(f'SPP Discount: {spp_discount}')

SPP Discount: -4007.668503886399


In [33]:
df_consume = df_infra2_agg[df_infra2_agg['bill_invoice_id']==1343605689]
credits = df_consume[df_consume['line_item_line_item_type']=='Credit']['line_item_unblended_cost'].sum()
print(f'Consume discount total: {credits}')

Consume discount total: -2843.3143923311


In [9]:
import pandas as pd
pd.set_option('display.max_rows', 100)
df_consume.groupby(by=['usage_account_name', 'product_product_name'])['line_item_unblended_cost'].sum().reset_index()

usage_account_name                            product_product_name  \
0            Bipp Care                               AWS Data Transfer   
1            Bipp Care                      AWS Key Management Service   
2            Bipp Care                                      AWS Lambda   
3            Bipp Care                                         AWS WAF   
4            Bipp Care  Amazon DocumentDB (with MongoDB compatibility)   
..                 ...                                             ...   
333             pakman                   Amazon Simple Storage Service   
334             pakman                                 Amazon Textract   
335             pakman                    Amazon Virtual Private Cloud   
336             pakman                                AmazonCloudWatch   
337             pakman                          Elastic Load Balancing   

     line_item_unblended_cost  
0                    0.340000  
1                    0.000216  
2                    0.304560  
3                   23.487219  
4                    0.002599  
..                        ...  
333                 31.906120  
334                  0.145000  
335                 83.565779  
336                126.369296  
337                 37.670411  

[338 rows x 3 columns]

In [24]:
df_infra2_agg.groupby(by=['bill_invoice_id', 'line_item_usage_account_id', 'usage_account_name', 'line_item_line_item_type']).agg(
    {'line_item_unblended_cost': 'sum', 'discount_spp_discount': 'sum'})

line_item_unblended_cost  \
bill_invoice_id line_item_usage_account_id usage_account_name                      line_item_line_item_type                             
1343605689      065716851408               SPI - SP Intervention - ENE Medical     Tax                                     106.610000   
                                                                                   Usage                                   866.206276   
                072038246234               SMIT                                    Tax                                    1126.850000   
                                                                                   Usage                                  9052.836375   
                095507049355               IHSL                                    Tax                                     106.600000   
                                                                                   Usage                                   865.950291   
                113082347426               Estre - Conta 2                         Tax                                      17.350000   
                                                                                   Usage                                   140.522065   
                167002523613               SPI - SP Intervention                   Tax                                     633.980000   
                                                                                   Usage                                  5457.117687   
                215028743710               Hospital Sabara                         Tax                                       0.010000   
                                                                                   Usage                                     0.071642   
                244835640994               GTM Chemicals                           Tax                                       0.000000   
                                                                                   Usage                                     0.002103   
                273762107511               Cordvida                                Tax                                     141.780000   
                                                                                   Usage                                  1220.446747   
                350427820123               Oliv-e                                  Tax                                       1.380000   
                                                                                   Usage                                    10.215014   
                368505959464               [DEV] OTIMO SOCIEDADE DE CREDITO DIRETO Tax                                     117.350000   
                                                                                   Usage                                   953.468511   
                415870527538               Hospital Santa Cruz                     Tax                                      43.120000   
                                                                                   Usage                                   350.234932   
                420065387364               pakman                                  Credit                                 -429.420000   
                                                                                   Tax                                     127.090000   
                                                                                   Usage                                  1514.789073   
                435585749363               EOS                                     Tax                                     136.180000   
                                                                                   Usage                                  1172.170658   
                446999942749               Ceramica Atlas                          Tax                                       0.470000   

In [15]:
df_consume_usage[df_consume_usage['line_item_unblended_cost']<0]

Empty DataFrame
Columns: [bill_invoice_id, bill_billing_entity, bill_invoicing_entity, bill_payer_account_id, payer_account_name, line_item_usage_account_id, usage_account_name, line_item_line_item_type, line_item_usage_start_date, line_item_usage_end_date, product_product_name, line_item_currency_code, line_item_unblended_cost]
Index: []

In [12]:
22056.207535710793 + 4007.41

26063.617535710793

In [6]:
df_consume[]['line_item_unblended_cost'].sum()

21596.333143379696

In [22]:
df_consume_credits = df_consume[df_consume['line_item_line_item_type']=='Credit']
print(f"Consume usage total: {df_consume_credits['line_item_unblended_cost'].sum()}")

Consume usage total: -2843.3143923311


In [19]:
df_consume_credits

Empty DataFrame
Columns: [bill_invoice_id, bill_billing_entity, bill_invoicing_entity, bill_payer_account_id, payer_account_name, line_item_usage_account_id, usage_account_name, line_item_line_item_type, line_item_usage_start_date, line_item_usage_end_date, product_product_name, line_item_currency_code, line_item_unblended_cost]
Index: []

In [31]:
df_consume = df_infra2_agg[df_infra2_agg['bill_invoice_id']==1343605689]
df_consume_total = df_consume[df_consume['line_item_line_item_type'].isin(['Tax', 'Usage', 'Fee'])]
print(f"Consume usage total: {df_consume_total['line_item_unblended_cost'].sum()}")

Consume usage total: 24439.647535710796
